# **BATCH NORMALIZATION**
---
---
### **HAND WRITTEN DIGIT PREDICTION**

---
---

## **Before Batch Normalization**

In [24]:
# import Libraries

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import time

In [5]:
# Load the Data
import tensorflow as tf


In [8]:

(X_train_full, y_train_full),(X_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

X_train_full = X_train_full/255.0
X_test = X_test/255.0

X_valid, X_train = X_train_full[:5000], X_train_full[5000:]
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

In [10]:
# Shape of Training, Validation and Test Data


print(X_train_full.shape)
print(X_train.shape)
print(X_valid.shape)
print(X_test.shape)

(60000, 28, 28)
(55000, 28, 28)
(5000, 28, 28)
(10000, 28, 28)


**Preparing Model**

In [15]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LeakyReLU, Flatten

In [14]:
model = Sequential()

In [17]:
# Input Layer
model.add(Flatten(input_shape = [28, 28]))

# Hidden Layer
model.add(Dense(300, kernel_initializer = "he_normal"))
model.add(LeakyReLU())

model.add(Dense(100, kernel_initializer = "he_normal"))
model.add(LeakyReLU())

# Output Layer
model.add(Dense(10, activation = "softmax"))


In [21]:
model.compile(loss = "sparse_categorical_crossentropy",
              optimizer = "SGD",
              metrics = ["accuracy"])

In [22]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense (Dense)               (None, 300)               235500    
                                                                 
 leaky_re_lu (LeakyReLU)     (None, 300)               0         
                                                                 
 dense_1 (Dense)             (None, 100)               30100     
                                                                 
 leaky_re_lu_1 (LeakyReLU)   (None, 100)               0         
                                                                 
 dense_2 (Dense)             (None, 10)                1010      
                                                                 
Total params: 266,610
Trainable params: 266,610
Non-trai

In [25]:
# Start Time :
start = time.time()

model.fit(X_train, y_train, epochs = 10, validation_data=[X_valid, y_valid])

# End Time:
end = time.time()

Epoch 1/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3450 - accuracy: 0.8773 - val_loss: 0.3570 - val_accuracy: 0.8742
Epoch 2/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3374 - accuracy: 0.8798 - val_loss: 0.3495 - val_accuracy: 0.8762
Epoch 3/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.3312 - accuracy: 0.8826 - val_loss: 0.3534 - val_accuracy: 0.8734
Epoch 4/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3248 - accuracy: 0.8848 - val_loss: 0.3564 - val_accuracy: 0.8714
Epoch 5/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3191 - accuracy: 0.8859 - val_loss: 0.3578 - val_accuracy: 0.8694
Epoch 6/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.3145 - accuracy: 0.8869 - val_loss: 0.3350 - val_accuracy: 0.8798
Epoch 7/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3098 - accuracy: 0.8883 - val_loss: 0.3473 - val_accuracy:

In [26]:
# Total time Taken

print(f"Runtime of our programe is {end-start}")

Runtime of our programe is 82.0340690612793


**CONCLUSION :**

Runtime of the Program : 82.03 Sec

We get the accuracy : 0.8732

---
---

## **After Applying Batch Normalization**

In [27]:
# import Batch Normalization

from tensorflow.keras.layers import BatchNormalization

In [30]:
model = Sequential()

In [31]:
# Input Layer
model.add(Flatten(input_shape = [28, 28]))
model.add(BatchNormalization())

# Hidden Layer
model.add(Dense(300, activation = "relu"))
model.add(BatchNormalization())

model.add(Dense(100, activation = "relu"))
model.add(BatchNormalization())

# Output Layer
model.add(Dense(10, activation = "softmax"))


In [34]:
model.compile(loss = "sparse_categorical_crossentropy",
              optimizer = "SGD",
              metrics = ["accuracy"])

In [32]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_2 (Flatten)         (None, 784)               0         
                                                                 
 batch_normalization_3 (Batc  (None, 784)              3136      
 hNormalization)                                                 
                                                                 
 dense_6 (Dense)             (None, 300)               235500    
                                                                 
 batch_normalization_4 (Batc  (None, 300)              1200      
 hNormalization)                                                 
                                                                 
 dense_7 (Dense)             (None, 100)               30100     
                                                                 
 batch_normalization_5 (Batc  (None, 100)             

In [35]:
# Start Time :
start = time.time()

model.fit(X_train, y_train, epochs = 10, validation_data=[X_valid, y_valid])

# End Time:
end = time.time()

Epoch 1/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.5380 - accuracy: 0.8128 - val_loss: 0.3787 - val_accuracy: 0.8654
Epoch 2/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.3948 - accuracy: 0.8602 - val_loss: 0.3503 - val_accuracy: 0.8766
Epoch 3/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3567 - accuracy: 0.8719 - val_loss: 0.3365 - val_accuracy: 0.8780
Epoch 4/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.3284 - accuracy: 0.8813 - val_loss: 0.3198 - val_accuracy: 0.8852
Epoch 5/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3098 - accuracy: 0.8879 - val_loss: 0.3215 - val_accuracy: 0.8864
Epoch 6/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2918 - accuracy: 0.8937 - val_loss: 0.3092 - val_accuracy: 0.8884
Epoch 7/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.2748 - accuracy: 0.8996 - val_loss: 0.3069 - val_accuracy:

In [36]:
# Total time Taken

print(f"Runtime of our programe is {end-start}")

Runtime of our programe is 82.69180011749268


**CONCLUSION :**

Runtime of the Program : 82.69 Sec

We get the accuracy : 0.8920

---
---
### **SUPER CONCLUSION :**

After Including Batch Normalization it increases the accuracy by 89% although it take approx same amount of time as earlier model.